titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/KubeApiServerIntegrator.java

```diff
@@ -519,37 +528,97 @@ private void podUpdated(V1Pod pod) {
+                    reasonCode = REASON_TASK_KILLED;
+                }
             } else if (phase.equalsIgnoreCase(FAILED)) {
-                publishContainerEvent(podName, Finished, REASON_FAILED, reasonMessage, executorDetails);
+                state = Finished;
+                reasonCode = REASON_FAILED;
+            } else {
+                titusRuntime.getCodeInvariants().inconsistent("Pod: %s has unknown phase mapping: %s", podName, phase);
+                return;
+            }
+
+            // if the new state is Finished and the pod had a container that ran then make sure that the StartInitiated
+            // and Started states were populated in the task state machine due to apiserver sending the latest event.
+            if (state == Finished) {
+                Optional<V1ContainerStateTerminated> terminatedContainerStatusOpt = findTerminatedContainerStatus(pod);
+                if (terminatedContainerStatusOpt.isPresent()) {
+                    V1ContainerStateTerminated containerStateTerminated = terminatedContainerStatusOpt.get();
+                    DateTime startedAt = containerStateTerminated.getStartedAt();
+                    if (startedAt != null) {
+                        long timestamp = startedAt.getMillis();
+                        Optional<TaskStatus> startInitiatedOpt = JobFunctions.findTaskStatus(task, StartInitiated);
+                        if (!startInitiatedOpt.isPresent()) {
+                            logger.debug("Publishing missing task status: StartInitiated for task: {}", task.getId());
+                            publishContainerEvent(podName, StartInitiated, TaskStatus.REASON_NORMAL, "",
+                                    Optional.empty(), timestamp);
+                        }
+                        Optional<TaskStatus> startedOpt = JobFunctions.findTaskStatus(task, Started);
+                        if (!startedOpt.isPresent()) {
+                            logger.debug("Publishing missing task status: Started for task: {}", task.getId());
+                            publishContainerEvent(podName, Started, TaskStatus.REASON_NORMAL, "",
+                                    Optional.empty(), timestamp);
+                        }
+                    }
+                }
             }
+
+            publishContainerEvent(podName, state, reasonCode, reasonMessage, executorDetails);
         } catch (Exception e) {
             logger.error("Unable to handle pod update: {} with error:", pod, e);
         }
     }
 
+    private Optional<V1ContainerStateTerminated> findTerminatedContainerStatus(V1Pod pod) {
+        List<V1ContainerStatus> containerStatuses = pod.getStatus().getContainerStatuses();
+        if (containerStatuses != null) {
+            for (V1ContainerStatus status : containerStatuses) {
+                V1ContainerState state = status.getState();
+                if (state != null) {
+                    return Optional.ofNullable(state.getTerminated());
```

Comment:
minor: never null, so you can use `Optional.of()`

Reply:
It is possible that it is null.

Reply:
my bad, I misread the null check just above